src/Spago/Build.hs

```diff
@@ -200,33 +202,117 @@ run
 run maybeModuleName buildOpts extraArgs = do
   Config.Config { alternateBackend } <- view (the @Config)
   let moduleName = fromMaybe (ModuleName "Main") maybeModuleName
-  runBackend alternateBackend moduleName Nothing "Running failed; " buildOpts extraArgs
+  runBackend alternateBackend Nothing moduleName Nothing "Running failed; " buildOpts extraArgs
 
 
+-- | Run the select module as a script: init, compile, and run the provided module
+script
+  :: (HasEnv env)
+  => Text
+  -> Maybe Text
+  -> [PackageName]
+  -> RIO env ()
+script modulePath tag packageDeps = do
+  logDebug "Running `spago script`"
+  absoluteModulePath <- fmap Text.pack (makeAbsolute (Text.unpack modulePath))
+  moduleContents <- readTextFile (pathFromText absoluteModulePath)
+
+  moduleName <- case parseModuleName moduleContents of
+    Nothing -> die [ "Failed to extract module name from source file" ]
+    Just modName -> pure modName
+
+  GlobalCache cacheDir _ <- view (the @GlobalCache)
+  currentDir <- Turtle.pwd
+  Temp.withTempDirectory cacheDir "spago-script-tmp" $ \dir -> do
+    let tmpDir = Text.pack dir
+    Turtle.cd (Turtle.fromText tmpDir)
+
+    config@Config { packageSet = PackageSet{..}, ..}
+      <- Packages.initProject NoForce Dhall.WithComments tag
+
+    let newConfig :: Config
+        newConfig = config
+          { Config.dependencies = dependencies <> packageDeps
+          , Config.configSourcePaths = []
+          }
+
+    Run.withInstallEnv' (Just newConfig) installAction
+
+    let runDirs :: RunDirectories
+        runDirs = RunDirectories tmpDir currentDir
+
+    Run.withBuildEnv' (Just newConfig) NoPsa (runAction absoluteModulePath moduleName runDirs)
+  where
+    installAction = do
+      deps <- Packages.getProjectDeps
+      Fetch.fetchPackages deps
+
+    runAction absoluteModulePath moduleName dirs = do
+      let
+        buildOpts = BuildOptions
+          { shouldWatch = BuildOnce
+          , shouldClear = NoClear
+          , allowIgnored = DoAllowIgnored
+          , sourcePaths = [ SourcePath absoluteModulePath ]
+          , withSourceMap = WithoutSrcMap
+          , noInstall = NoInstall
+          , pursArgs = []
+          , depsOnly = AllSources
+          , beforeCommands = []
+          , thenCommands = []
+          , elseCommands = []
+          }
+      runBackend Nothing (Just dirs) moduleName Nothing "error" buildOpts []
+
+    parseModuleName contents = do
+      Parse.PsModule name _ <- hush $ parse Parse.moduleDecl "" (encodeUtf8 contents)
+      hush $ ModuleName <$> decodeUtf8' name
+
+
+data RunDirectories = RunDirectories { tmpDir :: Text, executeDir :: FilePath }
```

Comment:
We've adjusted the `runBackend` function to accommodate the case in which the project is being executed in a temporary directory. In that case we need to track the temporary directory (as it contains the `run.js` file) as well as the directory where `spago script` was invoked (as this is the context where the script should be run).

For example: you run `spago script` on your desktop and expect to write a file to the current directory. We create a temporary directory, init a project, install dependencies, and compile. At this point we can't just execute the script as usual -- we need to change back to your desktop to execute the script. Otherwise you'll write your file into the temporary directory which will then be removed when the script exits.

You can see the effects of this change propagate through `runBackend`, which uses absolute paths when `RunDirectories` is provided and otherwise behaves as it did before.

Reply:
Sounds great. I think we can just use this in all cases - i.e. without the `Maybe` - and use absolute paths everywhere?